### PR TITLE
Add style guide to developer docs

### LINKED ISSUE #67: Developer Documentation: Style Guide
Markdown version of the style guide in the documentation.

### PR DESCRIPTION
**Pull request by oerc0122**
Add existing `coding_guide` from `Docs` to developer documentation, translated to Markdown and updated to reflect some more recent decisions.  

Fixes #67

### CODE REVIEW

File: docs/developer/style_guide.md [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+freely. If all code sticks to the specification at all times, then maintaining
+the code is much simpler.
+
+Note that the specification does not say anything about private routines or
+variables in any module. For this reason, great care must be taken when using
+such routines. Use the Implementation Guide as a reference for what such
+routines/variables do, and if you make any modifications, you must update the
+Implementation Guide and note the changes within the source code as well. The
+subprogram template is there to remind you to document all side-effects of each
+routine!
```
Comment (oerc0122) [inline docs/developer/style_guide.md]: Do we still want to refer to specification given the spec has not been updated for a while?

File: docs/developer/style_guide.md [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+
+```
+
+as this does not cause any copying of the array and allows the compiler to do
+bound-checking etc.
+
+The use of **assumed-shape arrays**, such as `int_array(:, :)` is less efficient
+as this has an explicit rank, but an implicit shape. This requires the use of
+functions such as `size(int_array,1)` and may prevent certain compiler
+optimisations. However, it does allow the compiler to do bound-checking.
```
Comment (oerc0122) [inline docs/developer/style_guide.md:155]: Are assumed-shape arrays now fully discouraged except in `Utility`?

File: docs/developer/style_guide.md [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+
+
+### Compiler/preprocessor directives
+
+These should be avoided wherever possible as they hinder portability. They should
+only be used in the `Utility` modules for machine specific code. The one exception
+to this rule, is the use of a single `#ifdef debug` block at the start
+of each routine to check for conditions of usage. For this reason, all module
+filenames should be given the extension `.F90` not `.f90` to enable
+preproccesing on the widest range of compilers available.
```
Comment (oerc0122) [inline docs/developer/style_guide.md]: No longer true?

File: docs/developer/style_guide.md [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+All code is to be contained within modules.
+
+
+### Usage
+
+Each module described in the specification is to be one source file. All
+routines given in the specification for this module are to be in this file.
+
+
+### Auxiliary modules
```
Comment (oerc0122) [inline docs/developer/style_guide.md:288]: No mention of `submodules` should be extended

File: docs/developer/style_guide.md [hunk 5]
```diff
@@ -0,0 +1,579 @@
...
+Emacs directives to set f90-mode, etc.
+
+
+### Case
+
+Use lower case for keywords and use capitalisation as given in the specification
+for all routines and variables mentioned therein. For private routines and variables
+use lower case in general. However, you may use upper case to differentiate
+words within a name, e.g. `userInputData` - as long as you are consistent
+throughout the module.
```
Comment (oerc0122) [inline docs/developer/style_guide.md:342]: Think we want to specify to use `lower_snake` case generally.

File: docs/developer/style_guide.md [hunk 6]
```diff
@@ -0,0 +1,579 @@
...
+
+Use spaces between keywords e.g. `end if` not `endif`.
+
+
+### Code cleanly
+
+Obviously, to ensure correctness and make it easier for others to maintain the
+code. If the obvious way to code a routine is inefficient, whilst the efficient
+way is obscure, then this is allowed as long as the routine is well documented,
+both internally and in this Implementation Guide, with a discussion as to the
```
Comment (oerc0122) [inline docs/developer/style_guide.md]: Do "implementation guide" references need to go? Similarly to spec, probably outdated

File: docs/developer/style_guide.md [hunk 7]
```diff
@@ -0,0 +1,579 @@
...
+design decisions made.
+
+
+### Be aware of vectorisation
+
+When coding, be aware that the code is often executed upon vector architectures.
```
Comment (oerc0122) [inline docs/developer/style_guide.md]: Can be removed? All machines these days (unless we're targetting castep for android and even then...)

File: docs/developer/style_guide.md [hunk 8]
```diff
@@ -0,0 +1,579 @@
...
+
+
+## Documentation
+
+
+### Internal
+
+Each module, based upon `ModuleTemplate.F90`, is to have a header block
+describing its overall purpose and function. Note that this header block also
+contains RCS directives to log changes to the code.
```
Comment (oerc0122) [inline docs/developer/style_guide.md]: They still do mostly, but should probably be removed

File: docs/developer/style_guide.md [hunk 9]
```diff
@@ -0,0 +1,579 @@
...
+repository. To this end a set of tests shall be provided, which must compile and
+execute correctly in the CI system. The purpose of these tests is to verify the correct
+operation of every specified routine in the module.
+
+
+### Reference data
+
+The module test program shall come with both appropriate input datafiles and
+reference output results. Correct operation of the test shall be verified by
+comparing the actual output against the reference output results.
```
Comment (oerc0122) [inline docs/developer/style_guide.md:445]: References to `testcode.py` and auto-generation thereof?

File: docs/developer/style_guide.md (hunk 1 above)
Comment (mijp1) [reply docs/developer/style_guide.md]: Yes, I think this is OK. One day we may move the spec to a more modern format ...

File: docs/developer/style_guide.md (hunk 2 above)
Comment (mijp1) [reply docs/developer/style_guide.md:155]: Not explicitly. I don't recall a conversation at CodeFest or elsewhere about changing this. Hence this is probably safe to leave in.

File: docs/developer/style_guide.md (hunk 3 above)
Comment (mijp1) [reply docs/developer/style_guide.md]: Agreed. This should say .F90 ONLY in Utility modules, and no preprocessor in Fundamental or Functional. We made many of the '#ifdef debug' tests unconditional on the grounds that there was little cost and lot of gain in having them wherever they were needed. This reduced the number of -D needed, and now only machine/OS ones in general (and CODATA versions etc).

File: docs/developer/style_guide.md (hunk 4 above)
Comment (mijp1) [reply docs/developer/style_guide.md:288]: Agreed. Submodules have now been recently added to our allowed code usage and this should reflect that.

File: docs/developer/style_guide.md (hunk 5 above)
Comment (mijp1) [reply docs/developer/style_guide.md:342]: It is optional. Some prefer it, others done. Hence if you WANT to use uppercase within a name then you can, but 'lower case in general' is CASTEP style.

File: docs/developer/style_guide.md (hunk 6 above)
Comment (mijp1) [reply docs/developer/style_guide.md]: Unfortunately, I think so. The 'Implementation Guide' was one of those really good ideas in theory that was just too onerous / unpopular / not as useful in reality as hoped; and consequently was not rigorously done; and consequently was not of a lot of use; and consequently was never kept up-to-date; and consequently should now be quietly dropped.

File: docs/developer/style_guide.md (hunk 7 above)
Comment (mijp1) [reply docs/developer/style_guide.md]: I think it is true that almost all modern architectures are vector to some level, and hence you SHOULD be aware of vectorisation when writing code. Hence code change 'often' to 'almost always' but the section should definitely stay.

File: docs/developer/style_guide.md (hunk 8 above)
Comment (mijp1) [reply docs/developer/style_guide.md]: Agreed. We moved away from RCS ~20 years ago!

File: docs/developer/style_guide.md (hunk 9 above)
Comment (mijp1) [reply docs/developer/style_guide.md:445]: Yes, good idea.

File: docs/developer/style_guide.md (hunk 2 above)
Comment (oerc0122) [reply docs/developer/style_guide.md:155]: Despite having an example 3 words down, my brain read it as assumed-size (`integer, dimension(*)`)

File: docs/developer/style_guide.md (hunk 3 above)
Comment (oerc0122) [reply docs/developer/style_guide.md]: Done and rephrased

File: docs/developer/style_guide.md (hunk 7 above)
Comment (oerc0122) [reply docs/developer/style_guide.md]: Done

File: docs/developer/style_guide.md (hunk 9 above)
Comment (oerc0122) [reply docs/developer/style_guide.md:445]: Done somewhat with (currently) dangling reference to files in #79

File: docs/developer/style_guide.md (hunk 6 above)
Comment (oerc0122) [reply docs/developer/style_guide.md]: Cleared references to the implementation guide.

File: docs/developer/style_guide.md (hunk 8 above)
Comment (oerc0122) [reply docs/developer/style_guide.md]: Done.

File: docs/developer/style_guide.md (hunk 4 above)
Comment (oerc0122) [reply docs/developer/style_guide.md:288]: Added paragraph to briefly introduce submodules.

**Review by mijp1 — APPROVED** [review]

**Pull request merged**